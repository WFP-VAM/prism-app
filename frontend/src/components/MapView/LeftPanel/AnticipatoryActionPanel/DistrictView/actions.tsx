import React from 'react';
import {
  MarkunreadOutlined,
  LocalDrink,
  School,
  AccountTree,
  Block,
  //   Diversity3 in mui 5 for social assistance
} from '@material-ui/icons';
import {
  faCow,
  faHandshake,
  faPeopleGroup,
  faPersonDigging,
  faPlateWheat,
  faSeedling,
  faSyringe,
} from '@fortawesome/free-solid-svg-icons';
import {
  FontAwesomeIcon,
  FontAwesomeIconProps,
} from '@fortawesome/react-fontawesome';

const FontAwesomeIconWrap = (props: FontAwesomeIconProps) => (
  <FontAwesomeIcon fontSize="1.5rem" {...props} />
);

interface Action {
  name: string;
  icon: any; // Placeholder type for the icon, adjust as necessary
}

// Simplified, reusable action items with full names
export const AActions = {
  warnings: {
    name: 'Dissemination of early warning messages',
    icon: <MarkunreadOutlined />,
  },
  seeds: {
    name: 'Distribution of seeds for re-sowing',
    icon: <FontAwesomeIconWrap icon={faSeedling} />,
  },
  seedsW2: {
    name: 'Distribution of seeds and vegetative material for the second season',
    // TODO - get 2 plant icon figma
    icon: <FontAwesomeIconWrap icon={faSeedling} />,
  },
  waterHoles: { name: 'Rehabilitation of boreholes', icon: <LocalDrink /> },
  multiUse: {
    name: 'Transform boreholes into multi-use systems',
    icon: <FontAwesomeIconWrap icon={faPersonDigging} />,
  },
  cattle: {
    name: 'Moving Cattle to places with water availability',
    icon: <FontAwesomeIconWrap icon={faCow} />,
  },
  vaccination: {
    name: 'Vaccination of birds against New Castle',
    icon: <FontAwesomeIconWrap icon={faSyringe} />,
  },
  foodAssistance: {
    name: 'Food Assistance',
    icon: <FontAwesomeIconWrap icon={faPlateWheat} />,
  },
  socialAssistance: {
    name: 'Social assistance',
    icon: <FontAwesomeIconWrap icon={faPeopleGroup} />,
  }, // replace with 'Diversity3'
  schoolLunch: { name: 'School lunches', icon: <School /> },
  procurement: {
    name: 'Launching tenders and other Procurement procedures',
    icon: <AccountTree />,
  },
  contracts: {
    name: 'Signing contracts and mobilization of assets',
    icon: <FontAwesomeIconWrap icon={faHandshake} />,
  },
  naMild: {
    name: 'No anticipatory action linked to Mild levels',
    icon: <Block />,
  },
};

// Define the map type
type ActionsMap = {
  [key: string]: Action[];
};

// Define the actions map
const actionsMap: ActionsMap = {
  readyModerateW1: [AActions.procurement],
  readyModerateW2: [AActions.procurement],
  readySevereW1: [AActions.procurement],
  readySevereW2: [AActions.procurement],
  setModerateW1: [
    AActions.contracts,
    AActions.warnings,
    AActions.seeds,
    AActions.waterHoles,
    AActions.multiUse,
  ],
  setModerateW2: [
    AActions.contracts,
    AActions.warnings,
    AActions.seeds,
    AActions.vaccination,
    AActions.socialAssistance,
  ],
  setSevereW1: [
    AActions.contracts,
    AActions.warnings,
    AActions.cattle,
    AActions.waterHoles,
    AActions.multiUse,
  ],
  setSevereW2: [
    AActions.contracts,
    AActions.warnings,
    AActions.foodAssistance,
    AActions.schoolLunch,
    AActions.socialAssistance,
  ],
};

// Function to get actions by state, severity, and window
export function getActionsByStateSeverityAndWindow(
  state: string,
  severity: string,
  window: string,
): Action[] {
  const key = `${state}${severity}${window}`;
  return actionsMap[key] || [];
}
