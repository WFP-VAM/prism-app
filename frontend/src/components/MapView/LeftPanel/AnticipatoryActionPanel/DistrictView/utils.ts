import {
  AnticipatoryActionDataRow,
  AnticipatoryActionState,
} from 'context/anticipatoryActionStateSlice/types';
import { AADataSeverityOrder } from '../utils';
import { TimelineRow } from '../Timeline/utils';

export function districtViewTransform(
  data: AnticipatoryActionDataRow[] | undefined,
  filters: AnticipatoryActionState['filters'],
) {
  if (!data) {
    return undefined;
  }
  const { categories: categoryFilters, selectedIndex } = filters;

  // Keep valid data only and switch SET phase to 'na' if the READY was not valid
  const validData = data
    .filter(
      x =>
        categoryFilters[x.category] &&
        (x.computedRow || x.isValid || x.wasReadyValid) &&
        (selectedIndex === '' || x.index === selectedIndex),
    )
    .map(x => {
      if (x.isValid) {
        return x;
      }
      return {
        ...x,
        phase: 'na',
      } as AnticipatoryActionDataRow;
    });

  const monthsMap = new Map<string, AnticipatoryActionDataRow[]>();
  validData.forEach(x => {
    const val = monthsMap.get(x.date);
    monthsMap.set(x.date, val ? [...val, x] : [x]);
  });

  // eslint-disable-next-line fp/no-mutating-methods
  const months = Array.from(monthsMap.keys()).sort();

  const topFiltered = months
    .map(date => {
      const dateData = monthsMap.get(date);
      if (!dateData) {
        // this should never happen
        throw new Error('Invalid Date');
      }

      if (dateData.filter(x => !x.computedRow).length === 0) {
        return [];
      }

      const setCategories = new Set(
        dateData.filter(x => x.phase === 'Set').map(x => x.category),
      );
      const ret = dateData.filter(
        x => x.phase === 'Set' || !setCategories.has(x.category),
      );

      return ret;
    })
    .flat();

  const sevMap = new Map<number, AnticipatoryActionDataRow[]>();
  topFiltered.forEach(x => {
    // return category data on one line by setting phase bonus to 0
    const sevVal = AADataSeverityOrder(x.category, x.phase, 0);
    const val = sevMap.get(sevVal);
    sevMap.set(sevVal, val ? [...val, x] : [x]);
  });

  const transformed = Object.fromEntries(sevMap);

  return {
    months: Object.fromEntries(
      [...new Set(topFiltered.map(x => x.date))].map(x => [
        x,
        new Date(x).toLocaleString('en-US', { month: 'short' }),
      ]),
    ),
    transformed,
  };
}

export function dateSorter(
  a: [string, AnticipatoryActionDataRow[] | TimelineRow],
  b: [string, AnticipatoryActionDataRow[] | TimelineRow],
) {
  if (a[0] > b[0]) {
    return -1;
  }
  if (a[0] < b[0]) {
    return 1;
  }
  return 0;
}
