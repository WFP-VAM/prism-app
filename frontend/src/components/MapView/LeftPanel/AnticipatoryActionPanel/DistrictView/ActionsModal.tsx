import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  IconButton,
  Typography,
  createStyles,
  makeStyles,
} from '@material-ui/core';
import { Cancel, Close } from '@material-ui/icons';
import { useSafeTranslation } from 'i18n';
import React from 'react';
import { black, cyanBlue } from 'muiTheme';
import { AActions } from './actions';

interface ActionsModalProps {
  open: boolean;
  onClose: () => void;
}

function ActionsModal({ open, onClose }: ActionsModalProps) {
  const { t } = useSafeTranslation();
  const classes = useStyles();

  return (
    <Dialog open={open}>
      <DialogTitle>
        <div className={classes.titleWrapper}>
          <Typography variant="h2">{t('Actions undertaken')}</Typography>
          <IconButton style={{ padding: 0 }} onClick={() => onClose()}>
            <Cancel />
          </IconButton>
        </div>
      </DialogTitle>
      <DialogContent dividers>
        <div className={classes.contentWrapper}>
          {Object.entries(AActions).map(([action, data]) => (
            <div key={action} className={classes.actionRow}>
              <div className={classes.actionIconWrapper}>{data.icon}</div>
              <Typography variant="h3">{data.name}</Typography>
            </div>
          ))}
        </div>
      </DialogContent>
      <DialogActions className={classes.dialogActionsWrapper}>
        <Button
          className={classes.dialogButton}
          variant="outlined"
          startIcon={<Close />}
          onClick={() => onClose()}
        >
          <Typography>{t('Close')}</Typography>
        </Button>
      </DialogActions>
    </Dialog>
  );
}

const useStyles = makeStyles(() =>
  createStyles({
    titleWrapper: {
      display: 'flex',
      flexDirection: 'row',
      flexWrap: 'nowrap',
      justifyContent: 'space-between',
    },
    contentWrapper: { display: 'flex', flexDirection: 'column', gap: '1rem' },
    actionRow: {
      display: 'flex',
      flexDirection: 'row',
      flexWrap: 'nowrap',
      gap: '1rem',
      color: 'black',
    },
    actionIconWrapper: { width: '2rem' },
    dialogActionsWrapper: { display: 'flex', justifyContent: 'center' },
    dialogButton: { borderColor: cyanBlue, color: black },
  }),
);

export default ActionsModal;
