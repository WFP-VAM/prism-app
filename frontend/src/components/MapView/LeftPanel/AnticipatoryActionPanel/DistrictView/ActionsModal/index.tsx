import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  IconButton,
  Typography,
  createStyles,
  makeStyles,
} from '@material-ui/core';
import { Cancel, Close } from '@material-ui/icons';
import { useSafeTranslation } from 'i18n';
import { black, cyanBlue } from 'muiTheme';
import { Action } from './actions';

interface ActionsModalProps {
  actions: Action[];
  open: boolean;
  onClose: () => void;
}

function ActionsModal({ open, onClose, actions }: ActionsModalProps) {
  const { t } = useSafeTranslation();
  const classes = useStyles();

  return (
    <Dialog open={open}>
      <DialogTitle>
        <div className={classes.titleWrapper}>
          <Typography variant="h2">{t('Recommended actions')}</Typography>
          <IconButton style={{ padding: 0 }} onClick={() => onClose()}>
            <Cancel />
          </IconButton>
        </div>
      </DialogTitle>
      <DialogContent dividers>
        <div className={classes.contentWrapper}>
          {actions.map(action => (
            <div key={action.name} className={classes.actionRow}>
              <div className={classes.actionIconWrapper}>{action.icon}</div>
              <Typography variant="h3">{t(action.name)}</Typography>
            </div>
          ))}
        </div>
      </DialogContent>
      <DialogActions className={classes.dialogActionsWrapper}>
        <Button
          className={classes.dialogButton}
          variant="outlined"
          startIcon={<Close />}
          onClick={() => onClose()}
        >
          <Typography>{t('Close')}</Typography>
        </Button>
      </DialogActions>
    </Dialog>
  );
}

const useStyles = makeStyles(() =>
  createStyles({
    titleWrapper: {
      display: 'flex',
      flexDirection: 'row',
      flexWrap: 'nowrap',
      justifyContent: 'space-between',
    },
    contentWrapper: { display: 'flex', flexDirection: 'column', gap: '1rem' },
    actionRow: {
      display: 'flex',
      flexDirection: 'row',
      flexWrap: 'nowrap',
      gap: '1rem',
      color: 'black',
    },
    actionIconWrapper: { width: '2rem' },
    dialogActionsWrapper: { display: 'flex', justifyContent: 'center' },
    dialogButton: { borderColor: cyanBlue, color: black },
  }),
);

export default ActionsModal;
