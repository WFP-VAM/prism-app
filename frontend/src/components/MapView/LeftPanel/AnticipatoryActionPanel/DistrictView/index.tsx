import { Typography, createStyles, makeStyles } from '@material-ui/core';
import { gray } from 'muiTheme';
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { setPanelSize } from 'context/leftPanelStateSlice';
import { PanelSize } from 'config/types';
import {
  AnticipatoryActionDataRow,
  AnticipatoryActionState,
} from 'context/anticipatoryActionStateSlice/types';
import { AAWindowKeys } from 'config/utils';
import {
  AADataSelector,
  AAFiltersSelector,
  AASelectedDistrictSelector,
} from 'context/anticipatoryActionStateSlice';
import { AADataSeverityOrder, getAAIcon } from '../utils';

function transform(
  data: AnticipatoryActionDataRow[] | undefined,
  filters: AnticipatoryActionState['filters'],
) {
  if (!data) {
    return undefined;
  }
  const { categories: categoryFilters, selectedIndex } = filters;

  const validData = data.filter(
    x =>
      (x.computedRow || x.isValid) &&
      categoryFilters[x.category] &&
      (selectedIndex === '' || x.index === selectedIndex),
  );

  const monthsMap = new Map<string, AnticipatoryActionDataRow[]>();
  validData.forEach(x => {
    const val = monthsMap.get(x.date);
    monthsMap.set(x.date, val ? [...val, x] : [x]);
  });

  // eslint-disable-next-line fp/no-mutating-methods
  const months = Array.from(monthsMap.keys()).sort();

  let prevMax: AnticipatoryActionDataRow | undefined;
  const topFiltered = months
    .map(date => {
      const dateData = monthsMap.get(date);
      if (!dateData) {
        // this should never happen
        throw new Error('Invalid Date');
      }

      if (dateData.filter(x => !x.computedRow).length === 0) {
        return [];
      }

      // eslint-disable-next-line fp/no-mutating-methods
      const sorted = dateData.sort((a, b) => {
        const aVal = AADataSeverityOrder(a.category, a.phase, 1);
        const bVal = AADataSeverityOrder(b.category, b.phase, 1);

        if (aVal > bVal) {
          return -1;
        }
        if (aVal < bVal) {
          return 1;
        }
        return 0;
      });
      if (prevMax === undefined) {
        // eslint-disable-next-line fp/no-mutation, prefer-destructuring
        prevMax = sorted[0];
        return [sorted[0]];
      }

      const ret = sorted.filter(
        x =>
          AADataSeverityOrder(x.category, x.phase, 1) >=
          AADataSeverityOrder(prevMax!.category, prevMax!.phase, 1),
      );

      // eslint-disable-next-line fp/no-mutation, prefer-destructuring
      prevMax = ret[0];

      return ret;
    })
    .flat();

  const sevMap = new Map<number, AnticipatoryActionDataRow[]>();
  topFiltered.forEach(x => {
    const sevVal = AADataSeverityOrder(x.category, x.phase, 1);
    const val = sevMap.get(sevVal);
    sevMap.set(sevVal, val ? [...val, x] : [x]);
  });

  const transformed = Object.fromEntries(sevMap);

  return {
    months: Object.fromEntries(
      [...new Set(topFiltered.map(x => x.date))].map(x => [
        x,
        new Date(x).toLocaleString('en-US', { month: 'long' }),
      ]),
    ),
    transformed,
  };
}

interface WindowColumnProps {
  win: typeof AAWindowKeys[number];
  transformed: ReturnType<typeof transform>;
  rowKeys: string[];
}

function WindowColumn({ win, transformed, rowKeys }: WindowColumnProps) {
  const classes = useWindowColumnStyles();

  const extendedTransformed = {
    ...Object.fromEntries(
      rowKeys.map(x => [x, [] as AnticipatoryActionDataRow[]]),
    ),
    ...(transformed?.transformed || {}),
  };

  if (!transformed) {
    return null;
  }

  return (
    <div className={classes.windowWrapper}>
      <div style={{ textAlign: 'center' }}>
        <Typography variant="h3" className={classes.headerText}>
          {win}
        </Typography>
      </div>
      <div className={classes.tableWrapper}>
        <div className={classes.rowWrapper}>
          {Object.entries(transformed?.months || {}).map(x => (
            <div key={x[0]} className={classes.column}>
              <Typography className={classes.monthText}>{x[1]}</Typography>
            </div>
          ))}
        </div>
        {
          // eslint-disable-next-line fp/no-mutating-methods
          Object.entries(extendedTransformed)
            .sort((a, b) => {
              if (a[0] > b[0]) {
                return -1;
              }
              if (a[0] < b[0]) {
                return 1;
              }
              return 0;
            })
            .map(x => (
              <div key={x[0]} className={classes.rowWrapper}>
                {Object.entries(transformed?.months || {}).map(y => {
                  const elem = x[1].find(z => z.date === y[0]);
                  if (!elem) {
                    return <div key={y[0]} className={classes.column} />;
                  }
                  return (
                    <div key={y[0]} className={classes.column}>
                      <div className={classes.iconWrapper}>
                        {getAAIcon(elem.category, elem.phase)}
                      </div>
                    </div>
                  );
                })}
              </div>
            ))
        }
      </div>
      <div className={classes.actionsWrapper}>
        <div style={{ textAlign: 'center', paddingTop: '0.5rem' }}>
          <Typography variant="h3">ACTIONS</Typography>
        </div>
        <div className={classes.actionBoxesWrapper}>
          {Object.keys(transformed.months).map(x => (
            <div id={String(x)} className={classes.actionBox} />
          ))}
        </div>
      </div>
    </div>
  );
}

const useWindowColumnStyles = makeStyles(() =>
  createStyles({
    windowWrapper: {
      display: 'flex',
      flexDirection: 'column',
      padding: '0.5rem 0.25rem',
    },
    tableWrapper: { display: 'flex', flexDirection: 'column', gap: '2px' },
    rowWrapper: {
      display: 'flex',
      flexDirection: 'row',
      minHeight: '3rem',
      background: 'white',
    },
    column: {
      width: '5.2rem',
      height: '3.2rem',
      padding: '0.1rem 0.25rem',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
    },
    headerText: {
      fontWeight: 'bold',
      textTransform: 'uppercase',
      height: '2rem',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
    },
    monthText: {
      background: gray,
      borderRadius: '4px',
      textAlign: 'center',
      textTransform: 'uppercase',
      lineHeight: '2rem',
      width: '100%',
    },
    iconWrapper: {
      width: '100%',
      height: '100%',
    },
    actionsWrapper: { display: 'flex', flexDirection: 'column', gap: '2px' },
    actionBoxesWrapper: {
      display: 'flex',
      flexDirection: 'row',
    },
    actionBox: {
      height: '6rem',
      width: '5.2rem',
      margin: '0.1rem 0.25rem',
      background: 'white',
      borderRadius: '4px',
    },
  }),
);

function DistrictView() {
  const dispatch = useDispatch();
  const classes = useDistrictViewStyles();
  const { selectedWindow } = useSelector(AAFiltersSelector);
  const rawAAData = useSelector(AADataSelector);
  const aaFilters = useSelector(AAFiltersSelector);
  const selectedDistrict = useSelector(AASelectedDistrictSelector);

  const windows = selectedWindow === 'All' ? AAWindowKeys : [selectedWindow];
  const transformed = windows.map(x =>
    transform(rawAAData[x][selectedDistrict], aaFilters),
  );
  const rowKeys = transformed
    .map(x => Object.keys(x?.transformed || {}))
    .flat();

  React.useEffect(() => {
    dispatch(setPanelSize(PanelSize.undef));
  }, [dispatch]);

  return (
    <div className={classes.root}>
      <div className={classes.districtViewWrapper}>
        {windows.map((win, index) => (
          <WindowColumn
            key={win}
            win={win}
            transformed={transformed[index]}
            rowKeys={rowKeys}
          />
        ))}
      </div>
    </div>
  );
}

const useDistrictViewStyles = makeStyles(() =>
  createStyles({
    root: {
      display: 'flex',
      flexDirection: 'column',
      width: '100%',
      background: gray,
      overflow: 'scroll',
    },
    districtViewWrapper: {
      display: 'flex',
      flexDirection: 'row',
      width: '100%',
      background: gray,
      overflow: 'scroll',
      justifyContent: 'center',
    },
  }),
);

export default DistrictView;
