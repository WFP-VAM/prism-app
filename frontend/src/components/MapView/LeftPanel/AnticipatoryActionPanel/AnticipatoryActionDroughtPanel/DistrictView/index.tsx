import {
  Button,
  Tooltip,
  Typography,
  createStyles,
  makeStyles,
} from '@material-ui/core';
import { lightGrey } from 'muiTheme';
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  AAView,
  AnticipatoryActionDataRow,
} from 'context/anticipatoryAction/AADroughtStateSlice/types';
import { AAWindowKeys } from 'config/utils';
import {
  AADataSelector,
  AAFiltersSelector,
  AASelectedDistrictSelector,
  setAASelectedDistrict,
  setAAView,
} from 'context/anticipatoryAction/AADroughtStateSlice';
import { useSafeTranslation } from 'i18n';
import { ClearAll, Equalizer, Reply } from '@material-ui/icons';
import { AADataSeverityOrder, getAAIcon } from '../utils';
import { useAACommonStyles } from '../../utils';
import {
  Action,
  getActionsByPhaseCategoryAndWindow,
} from './ActionsModal/actions';
import ActionsModal from './ActionsModal';
import { dateSorter, districtViewTransform } from './utils';

interface WindowColumnProps {
  win: (typeof AAWindowKeys)[number];
  transformed: ReturnType<typeof districtViewTransform>;
  rowKeys: string[];
  openActionsDialog: () => void;
  setModalActions: React.Dispatch<React.SetStateAction<Action[]>>;
}

function WindowColumn({
  win,
  transformed,
  rowKeys,
  openActionsDialog,
  setModalActions,
}: WindowColumnProps) {
  const classes = useWindowColumnStyles();
  const { t } = useSafeTranslation();

  const extendedTransformed = {
    ...Object.fromEntries(
      rowKeys.map(x => [x, [] as AnticipatoryActionDataRow[]]),
    ),
    ...(transformed?.transformed || {}),
  };

  const allEntries = Object.values(transformed?.transformed || {}).flat();

  if (!transformed) {
    return null;
  }
  const hasWindowData = Object.keys(transformed.months).length > 0;

  return (
    <div className={classes.windowWrapper}>
      <div style={{ textAlign: 'center' }}>
        <Typography variant="h4" className={classes.headerText}>
          {t(win)}
        </Typography>
      </div>
      {!hasWindowData && (
        <Typography variant="h3" className={classes.noDataText}>
          {t('No data')}
        </Typography>
      )}
      {hasWindowData && (
        <div className={classes.tableWrapper}>
          <div className={classes.headRowWrapper}>
            {Object.entries(transformed?.months || {}).map(x => (
              <div key={x[0]} className={classes.headColumn}>
                <Typography className={classes.monthText}>{t(x[1])}</Typography>
              </div>
            ))}
          </div>
          {
            // eslint-disable-next-line fp/no-mutating-methods
            Object.entries(extendedTransformed)
              .sort(dateSorter)
              .map(x => (
                <div key={x[0]} className={classes.rowWrapper}>
                  {Object.entries(transformed?.months || {}).map(y => {
                    const elem = x[1].find(z => z.date === y[0]);
                    if (!elem) {
                      return <div key={y[0]} className={classes.column} />;
                    }
                    return (
                      <div key={y[0]} className={classes.column}>
                        <div className={classes.iconWrapper}>
                          {getAAIcon(elem.category, elem.phase)}
                        </div>
                      </div>
                    );
                  })}
                </div>
              ))
          }
        </div>
      )}
      {hasWindowData && (
        <div className={classes.actionsWrapper}>
          <div style={{ textAlign: 'center' }}>
            <Typography variant="h3">{t('ACTIONS')}</Typography>
          </div>
          <div className={classes.actionBoxesWrapper}>
            {Object.keys(transformed.months).map(x => {
              const dateData = allEntries.filter(y => y.date === x);
              const elem = dateData.reduce(
                (max, curr) =>
                  AADataSeverityOrder(max.category, max.phase) >
                  AADataSeverityOrder(curr.category, curr.phase)
                    ? max
                    : curr,
                dateData[0],
              );
              const actions = getActionsByPhaseCategoryAndWindow(
                elem.phase,
                elem.category,
                elem.window,
              );
              return (
                <button
                  key={x}
                  type="button"
                  id={String(x)}
                  className={classes.actionBox}
                  style={{
                    justifyContent: actions.length <= 2 ? 'center' : undefined,
                    cursor: actions.length > 0 ? 'pointer' : undefined,
                  }}
                  onClick={() => {
                    if (actions.length === 0) {
                      return;
                    }
                    openActionsDialog();
                    setModalActions(actions);
                  }}
                >
                  {actions.map(action => (
                    // wrapping in div to show tooltip with FontAwesomeIcons
                    <Tooltip
                      key={action.name}
                      title={t(action.name) || action.name}
                    >
                      <div>{action.icon}</div>
                    </Tooltip>
                  ))}
                </button>
              );
            })}
          </div>
        </div>
      )}
    </div>
  );
}

const useWindowColumnStyles = makeStyles(() =>
  createStyles({
    windowWrapper: {
      display: 'flex',
      flexDirection: 'column',
      padding: '0rem 0.25rem',
    },
    tableWrapper: { display: 'flex', flexDirection: 'column', gap: '2px' },
    rowWrapper: {
      display: 'flex',
      flexDirection: 'row',
      minHeight: '3rem',
      background: 'white',
    },
    headRowWrapper: {
      display: 'flex',
      flexDirection: 'row',
      minHeight: '2.5rem',
      background: 'white',
    },
    headColumn: {
      width: '4.6rem',
      padding: '0.1rem 0.25rem',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
    },
    column: {
      width: '4.6rem',
      height: '3.2rem',
      padding: '0.1rem 0.25rem',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
    },
    headerText: {
      fontWeight: 'bold',
      textTransform: 'uppercase',
      height: '2rem',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
    },
    monthText: {
      background: lightGrey,
      borderRadius: '4px',
      textAlign: 'center',
      textTransform: 'uppercase',
      lineHeight: '2rem',
      width: '100%',
    },
    iconWrapper: {
      width: '100%',
      height: '100%',
    },
    actionsWrapper: {
      display: 'flex',
      flexDirection: 'column',
      gap: '2px',
      padding: '0.5rem 0',
    },
    actionBoxesWrapper: {
      display: 'flex',
      flexDirection: 'row',
    },
    actionBox: {
      height: '6.2rem',
      width: '4.6rem',
      margin: '0.1rem 0.25rem',
      background: 'white',
      borderRadius: '4px',
      color: 'black',
      display: 'flex',
      flexWrap: 'wrap',
      flexDirection: 'column',
      alignItems: 'center',
      gap: '0.5rem',
      paddingTop: '0.8rem',
      paddingBottom: '0.8rem',
      border: 'none',
    },
    noDataText: {
      margin: 'auto',
    },
  }),
);

interface DistrictViewProps {
  dialogs: {
    text: string;
    onclick: () => void;
  }[];
}

function DistrictView({ dialogs }: DistrictViewProps) {
  const dispatch = useDispatch();
  const { t } = useSafeTranslation();
  const classes = useDistrictViewStyles();
  const commonClasses = useAACommonStyles();
  const { selectedWindow } = useSelector(AAFiltersSelector);
  const rawAAData = useSelector(AADataSelector);
  const aaFilters = useSelector(AAFiltersSelector);
  const selectedDistrict = useSelector(AASelectedDistrictSelector);
  const [actionsModalOpen, setActionsModalOpen] =
    React.useState<boolean>(false);
  const [modalActions, setModalActions] = React.useState<Action[]>([]);

  const districtButtons = [
    {
      icon: ClearAll,
      text: 'Timeline',
      onClick: () => dispatch(setAAView(AAView.Timeline)),
    },
    {
      icon: Equalizer,
      text: 'Forecast',
      onClick: () => dispatch(setAAView(AAView.Forecast)),
    },
    {
      icon: Reply,
      text: 'Summary',
      onClick: () =>
        dispatch(setAAView(AAView.Home)) && dispatch(setAASelectedDistrict('')),
    },
  ];

  const windows = selectedWindow === 'All' ? AAWindowKeys : [selectedWindow];
  const transformed = windows.map(x =>
    districtViewTransform(rawAAData[x][selectedDistrict], aaFilters),
  );
  const rowKeys = transformed
    .map(x => Object.keys(x?.transformed || {}))
    .flat();

  return (
    <>
      <div className={classes.root}>
        <ActionsModal
          open={actionsModalOpen}
          onClose={() => setActionsModalOpen(false)}
          actions={modalActions}
        />
        <div className={classes.districtViewWrapper}>
          {windows.map((win, index) => (
            <WindowColumn
              key={win}
              win={win}
              transformed={transformed[index]}
              rowKeys={rowKeys}
              openActionsDialog={() => setActionsModalOpen(true)}
              setModalActions={setModalActions}
            />
          ))}
        </div>
      </div>
      <div className={commonClasses.footerWrapper}>
        <div className={commonClasses.footerActionsWrapper}>
          {districtButtons.map(x => (
            <Button
              key={x.text}
              className={commonClasses.footerButton}
              variant="outlined"
              fullWidth
              onClick={x.onClick}
              startIcon={<x.icon />}
            >
              <Typography>{t(x.text)}</Typography>
            </Button>
          ))}
        </div>
        <div className={commonClasses.footerDialogsWrapper}>
          {dialogs.map(dialog => (
            <Typography
              key={dialog.text}
              className={commonClasses.footerDialog}
              component="button"
              onClick={() => dialog.onclick()}
            >
              {t(dialog.text)}
            </Typography>
          ))}
        </div>
      </div>
    </>
  );
}

const useDistrictViewStyles = makeStyles(() =>
  createStyles({
    root: {
      display: 'flex',
      flexDirection: 'column',
      width: '100%',
      background: lightGrey,
      overflow: 'scroll',
    },
    districtViewWrapper: {
      display: 'flex',
      flexDirection: 'row',
      width: '100%',
      background: lightGrey,
      overflow: 'scroll',
      justifyContent: 'space-around',
    },
  }),
);

export default DistrictView;
