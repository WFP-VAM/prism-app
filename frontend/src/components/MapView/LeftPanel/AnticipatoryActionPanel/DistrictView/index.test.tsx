import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import DistrictView from '.';
import { mockAAData } from '../test.utils';

const mockStore = configureStore([]);

const store = mockStore({
  mapState: {
    layers: [],
    dateRange: { startDate: 1701432000000 },
    maplibreMap: () => {},
    errors: [],
    layersData: [],
    loadingLayerIds: [],
    boundaryRelationData: {},
  },
  serverState: { availableDates: {}, loading: false },
  anticipatoryActionState: {
    filters: {
      selectedWindow: 'Window 2',
      categories: {
        Severo: true,
        Moderado: true,
        Leve: true,
      },
    },
    monitoredDistricts: ['Caia'],
    data: mockAAData,
  },
});

test('renders as expected', () => {
  const { container } = render(
    <BrowserRouter>
      <Provider store={store}>
        <DistrictView selectedDistrict="Caia" />
      </Provider>
    </BrowserRouter>,
  );
  expect(container).toMatchSnapshot();
});
