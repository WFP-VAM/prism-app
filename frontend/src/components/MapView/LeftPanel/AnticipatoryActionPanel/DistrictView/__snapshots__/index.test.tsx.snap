// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <div
    class="makeStyles-root-1"
  >
    <div
      class="makeStyles-districtViewWrapper-2"
    />
  </div>
</div>
`;
